cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/autoinstall/HubPackage.java

```diff
@@ -0,0 +1,254 @@
+    this.name = name;
+    this.version = version;
+    this.artifactVersion = new ArtifactVersion(version);
+    this.description = description;
+    this.label = label;
+    this.author = author;
+    this.org = org;
+    this.cdapVersion = cdapVersion;
+    this.versionRange = ArtifactVersionRange.parse(cdapVersion);
+    this.created = created;
+    this.beta = beta;
+    this.categories = categories;
+    this.paid = paid;
+    this.paidLink = paidLink;
+  }
+
+  private String getCdapVersion() {
+    return cdapVersion;
+  }
+
+  public ArtifactVersion getArtifactVersion() {
+    if (artifactVersion == null) {
+      // artifactVersion will be null if the class was deserialized from json.
+      artifactVersion = new ArtifactVersion(version);
+
+    }
+    return artifactVersion;
+  }
+
+  public boolean versionIsInRange(String version) {
+    if (versionRange == null) {
+      // versionRange will be null if the class was deserialized from json.
+      if (cdapVersion == null) {
+        return false;
+      }
+      try {
+        versionRange = ArtifactVersionRange.parse(cdapVersion);
+      } catch (InvalidArtifactRangeException e) {
+        LOG.error(String.format("%s has invalid artifact range", name), e);
+        return false;
+      }
+    }
+    return versionRange.versionIsInRange(new ArtifactVersion(version));
+  }
+  public void installPlugin(URL url, ArtifactRepository artifactRepository, File tmpDir) throws Exception {
+    // Deserialize spec.json
+    URL specURL = new URL(url, "/v2/packages/" + name + "/" + version + "/spec.json");
+    Spec spec = GSON.fromJson(HttpClients.doGetAsString(specURL), Spec.class);
+    for (Spec.Action action: spec.getActions()) {
+      // Get plugin jar and json names under one_step_deploy_plugin action in the spec
+      // See https://cdap.atlassian.net/wiki/spaces/DOCS/pages/554401840/Hub+API?src=search#one_step_deploy_plugin
+      if (!action.getType().equals("one_step_deploy_plugin")) {
+        continue;
+      }
+
+      String configFilename = action.getConfigFilename();
+      if (configFilename == null) {
+        LOG.error("Config not found for plugin {}", name);
+        continue;
+      }
+      URL configURL = new URL(url, Joiner.on("/").join(Arrays.asList("/v2/packages", name, version, configFilename)));
+      // Download plugin json from hub
+      JsonObject jsonObj = GSON.fromJson(HttpClients.doGetAsString(configURL), JsonObject.class);
+      List<String> parents = GSON.fromJson(jsonObj.get("parents").toString(), new TypeToken<ArrayList<String>>() {
+      }.getType());
+
+      String jarName = action.getJarName();
+      if (jarName == null) {
+        LOG.error("Jar not found for plugin {}", name);
+        continue;
+      }
+      URL jarURL = new URL(url, Joiner.on("/").join(Arrays.asList("/v2/packages", name, version, jarName)));
+      // Download plugin jar from hub
+      byte[] jar = HttpClients.doGetAsBytes(jarURL);
```

Comment:
Why download the bytes in memory? You can write out to the tmp file directly.

Reply:
fixed to use streaming API